### DISCUSSION TITLE
STM32WL series radio power output

### DISCUSSION THREAD

**Discussion by Martdur** [zephyrproject-rtos/zephyr#75312] [Q&A]
As discussed here [56833](https://github.com/zephyrproject-rtos/zephyr/pull/56833) the support for STM32WL power amplifier configuration now allows to select between `rfo-hp` and `rfo-lp`.
But a comment says that low power output is not currently supported by the driver : 
```
` /* High-power output is selected as a consequence of using
 * tx/rx-enable-gpio to control FE_CTRL1 and FE_CTRL2. Low-power
 * output would require both FE_CTRL1 and FE_CTRL2 to be high,
 * which is not currently supported by the driver.
 */`
```
So setting `power-amplifier-output = "rfo-lp";` won't work as I understand. 

Is there a way to modify the device tree parameter to make this work or maybe there is something I'm missing.

Do you have any input to this @knthm ? 
And thanks by the way for the work in this PR !

**Comment by erwango** [comment]
As @knthm is not active anymore, maybe @martinjaeger you would be able to answer as an active (at least former) STM32WL user ?

**Comment by ndrs-pst** [comment]
@Martdur Alongside setting `power-amplifier-output = "rfo-lp";` 
due to the RF Front-end schematic 
(https://www.st.com/resource/en/schematic_pack/mb1389-wl55jc-highband-d04_schematic.pdf), 
you may need to modify it to pass ~~`1` to `sx126x_set_rx_enable(1);`~~ `0` to `sx126x_set_tx_enable(0);` instead.

https://github.com/zephyrproject-rtos/zephyr/blob/6700a154be3ca29be50fb4f1e35d899de69be5ca/drivers/lora/sx126x.c#L286-L289

**Comment by ndrs-pst** [comment]
According to the [stm32wlxx-nucleo-bsp](https://github.com/STMicroelectronics/stm32wlxx-nucleo-bsp)

[stm32wlxx_nucleo_radio.c#L105-L146](https://github.com/STMicroelectronics/stm32wlxx-nucleo-bsp/blob/aafd038049d0aca9e95dc9c8920722f5dce31a91/stm32wlxx_nucleo_radio.c#L105-L146)

~~I think the followings should be GPIO_ACTIVE_HIGH~~
```
tx-enable-gpios = <&gpioc 4 GPIO_ACTIVE_LOW>; /* FE_CTRL1 */
rx-enable-gpios = <&gpioc 5 GPIO_ACTIVE_LOW>; /* FE_CTRL2 */
```

**Reply by martinjaeger** [reply]
The existing devicetree spec looks correct to me, see excerpt from Nucleo schematic:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/17674105/e4b95a2f-c8be-4a25-88ba-c0310fb1c1c2)

**Reply by ndrs-pst** [reply]
Yeah, you're right. Thanks.

**Comment by martinjaeger** [comment]
@Martdur The `power-amplifier-output = "rfo-lp";` should be supported in general. I'm using the Olimex board in some projects, which [has the LP amplifier configured](https://github.com/zephyrproject-rtos/zephyr/blob/a6e672f9b2368e6d011da189afd462fec599fd0d/dts/arm/olimex/bb-stm32wl.dtsi). It's just not supported on the Nucleo board because it would require setting of 3 different pins to the correct levels (see table posted above) and we can currently only configure the rx-enable and tx-enable pins.

**Reply by Martdur** [reply]
Thanks for all the information you've bring.
So for the nucelo 32wl55jc I can configure 
```
tx-enable-gpios = <&gpioc 4 GPIO_ACTIVE_HIGH>; /* FE_CTRL1 */
rx-enable-gpios = <&gpioc 5 GPIO_ACTIVE_LOW>; /* FE_CTRL2 */
```
with `power-amplifier-output = "rfo-lp";` and it should work specifically for the lp mode ?

**Reply by Martdur** [reply]
And, Can the use of `power-amplifier-output = "rfo-lp";` on nucleo stm32wl55jc without any changes can damage the board ? 
I think I've broke some of mine with this kind of settings.
If so it'd be nice to remove this property for this board as long as the feature is not implemented.

**Reply by ndrs-pst** [reply]
> Thanks for all the information you've bring. So for the nucelo 32wl55jc I can configure
> 
> ```
> tx-enable-gpios = <&gpioc 4 GPIO_ACTIVE_HIGH>; /* FE_CTRL1 */
> rx-enable-gpios = <&gpioc 5 GPIO_ACTIVE_LOW>; /* FE_CTRL2 */
> ```
> 
> with `power-amplifier-output = "rfo-lp";` and it should work specifically for the lp mode ?

Instead of changing to `sx126x_set_tx_enable(0);` only case MODE_TX: changing to GPIO_ACTIVE_HIGH will effect other switch case which will bring undesired result of pin control to `RF SP3T Switch`
https://github.com/zephyrproject-rtos/zephyr/blob/6700a154be3ca29be50fb4f1e35d899de69be5ca/drivers/lora/sx126x.c#L291-L306

**Reply by martinjaeger** [reply]
> Thanks for all the information you've bring. So for the nucelo 32wl55jc I can configure
> 
> ```
> tx-enable-gpios = <&gpioc 4 GPIO_ACTIVE_HIGH>; /* FE_CTRL1 */
> rx-enable-gpios = <&gpioc 5 GPIO_ACTIVE_LOW>; /* FE_CTRL2 */
> ```
> 
> with `power-amplifier-output = "rfo-lp";` and it should work specifically for the lp mode ?

I don't think this will work, as the driver will switch between RX and TX mode, so the pins need to be mutually exclusive. Take a look at the truth table:

|   | FE_CTRL1 | FE_CTRL2
|--|--|-----|
| TX mode |High | High |
| RX mode | Low | Low |

The RX mode is **not a valid combination** from the above table from the Nucleo schematic. Not sure if it could destroy the hardware.

**Reply by Martdur** [reply]
Ok thanks a lot !
I don't know if I close the discussion as it can be a feature to be implemented later on (maybe by adding a specific implementation of the driver for this kind of board).
I know that this feature was implemented in mbdedos and in stmCube.
What do you think ?

**Reply by martinjaeger** [reply]
I'd suggest to open an enhancement issue describing the results of the discussion. That way it's easier to find in the future. (at least I don't usually look into the discussions unless I'm pinged by someone)

**Comment by ndrs-pst** [comment]
@Martdur, @martinjaeger I'll try to open a PR (as an enhancement) within today. Please wait a moment. 😃

---
Here we go!!! https://github.com/zephyrproject-rtos/zephyr/pull/75445
